content/blog/2018/appswitch/appswitch.md

```diff
@@ -0,0 +1,81 @@
+
+The central AppSwitch construct that achieves the decoupling between services objects and their identities is _service reference_ (_reference_, for short).  AppSwitch implements service references based on the API instrumentation mechanism outlined above.  A service reference consists of an IP:port pair (and optionally a DNS name) and a label-selector that selects the service represented by the reference and the clients to which this reference applies.  A reference supports a few key properties.  (1) It can be named independently of the name of the object it refers to.  That is, a service may be listening on an IP and port but a reference allows that service to be reached on any other IP and port chosen by the user.  This is what allows AppSwitch to run traditional applications captured from their source environments with static configurations to run on Kubernetes by providing them with arbirary IP addresses and ports regardless of underlying network environment at the target.  (2) It remains unchanged even if the location of the target service changes.  A reference automatically redirects itself as its label-selector now resolves to the new instance of the service (3) Most important for this discussion, a reference remains valid even as the target service is coming up.
+
+To facilitate discovering services that can be accessed through service references, AppSwitch provides a _service registry_.  The registry is auto-curated and automatically kept up to date as services come and go across the cluster based on the network API that AppSwitch tracks.  Each entry in the registry consists of the IP and port where the respective service is bound.  Along with that, it includes a set of labels indicating the application to which this service belongs, the IP and port that the application passed through the socket API when creating the service, the IP and port where AppSwitch actually bound the service on the underlying host on behalf of the application etc.  In addition, applications created under AppSwitch carry a set of labels passed by the user that describe the application together with a few default system labels indicating the user that created the application and the host where the application is running etc.  These labels are all available to be expressed in the label-selector carried by a service reference.  A service in the registry can be made accessible to clients by creating a service reference.  A client would then be able to reach the service at the reference’s name (IP:port).  Now let’s look at how AppSwitch guarantees that the reference remains valid even when the target service has not yet come up.
+
+### Nonblocking Requests
```

Comment:
For this and other headers, please only capitalize the first letter and keep the rest in lowercase (except for acronyms and proper names)